tsdb/db_test.go

```diff
@@ -3489,3 +3490,225 @@ func TestDBPanicOnMmappingHeadChunk(t *testing.T) {
+	// and a changed metadata entry to the second series.
+	m4 := metadata.Metadata{Type: "counter", Unit: "unit_4", Help: "help_4"}
+	m5 := metadata.Metadata{Type: "counter", Unit: "unit_5", Help: "help_5"}
+	app = db.Appender(ctx)
+	app.UpdateMetadata(0, s1, m1)
+	app.UpdateMetadata(0, s4, m4)
+	app.UpdateMetadata(0, s2, m5)
+	app.Commit()
+
+	// Read the WAL to see if the disk storage format is correct.
+	recs := readTestWAL(t, path.Join(db.Dir(), "wal"))
+	var gotMetadataBlocks [][]record.RefMetadata
+	for _, rec := range recs {
+		if mr, ok := rec.([]record.RefMetadata); ok {
+			gotMetadataBlocks = append(gotMetadataBlocks, mr)
+		}
+	}
+
+	expectedMetadata := []record.RefMetadata{
+		{Ref: 1, Type: record.GetMetricType(m1.Type), Unit: m1.Unit, Help: m1.Help},
+		{Ref: 2, Type: record.GetMetricType(m2.Type), Unit: m2.Unit, Help: m2.Help},
+		{Ref: 3, Type: record.GetMetricType(m3.Type), Unit: m3.Unit, Help: m3.Help},
+		{Ref: 4, Type: record.GetMetricType(m4.Type), Unit: m4.Unit, Help: m4.Help},
+		{Ref: 2, Type: record.GetMetricType(m5.Type), Unit: m5.Unit, Help: m5.Help},
+	}
+	require.Len(t, gotMetadataBlocks, 2)
+	require.Equal(t, expectedMetadata[:3], gotMetadataBlocks[0])
+	require.Equal(t, expectedMetadata[3:], gotMetadataBlocks[1])
+}
+
+func TestMetadataCheckpointingOnlyKeepsLatestEntry(t *testing.T) {
+	ctx := context.Background()
+	numSamples := 10000
+	hb, w := newTestHead(t, int64(numSamples)*10, false)
+
+	// Add some series so we can append metadata to them.
+	app := hb.Appender(ctx)
+	s1 := labels.FromStrings("a", "b")
+	s2 := labels.FromStrings("c", "d")
+	s3 := labels.FromStrings("e", "f")
+	s4 := labels.FromStrings("g", "h")
+	app.Append(0, s1, 0, 0)
+	app.Append(0, s2, 0, 0)
+	app.Append(0, s3, 0, 0)
+	app.Append(0, s4, 0, 0)
+	app.Commit()
+
+	// Add a first round of metadata to the first three series.
+	// Re-take the Appender, as the previous Commit will have it closed.
+	app = hb.Appender(ctx)
+	m1 := metadata.Metadata{Type: "gauge", Unit: "unit_1", Help: "help_1"}
+	m2 := metadata.Metadata{Type: "gauge", Unit: "unit_2", Help: "help_2"}
+	m3 := metadata.Metadata{Type: "gauge", Unit: "unit_3", Help: "help_3"}
+	m4 := metadata.Metadata{Type: "gauge", Unit: "unit_4", Help: "help_4"}
+	app.UpdateMetadata(0, s1, m1)
+	app.UpdateMetadata(0, s2, m2)
+	app.UpdateMetadata(0, s3, m3)
+	app.UpdateMetadata(0, s4, m4)
+	app.Commit()
+
+	// Update metadata for first series.
+	app = hb.Appender(ctx)
+	m5 := metadata.Metadata{Type: "counter", Unit: "unit_5", Help: "help_5"}
+	app.UpdateMetadata(0, s1, m5)
+	app.Commit()
+
+	// Switch back-and-forth metadata for second series.
+	// Since it ended on a new metadata record, we expect a single new entry.
+	app = hb.Appender(ctx)
+	m6 := metadata.Metadata{Type: "counter", Unit: "unit_6", Help: "help_6"}
+	app.UpdateMetadata(0, s2, m6)
+	app.UpdateMetadata(0, s2, m2)
+	app.UpdateMetadata(0, s2, m6)
+	app.UpdateMetadata(0, s2, m2)
+	app.UpdateMetadata(0, s2, m6)
```

Comment:
// app = ...
// update m6
// commit()

// app = ...
// update m2
// commit()

// app = ...
// update m6
// commit()

...and so on. One appender for each. If you want to know why the current state is a problem - try ending it at m2 by removing the last line and see no test failure though it should have failed.